README.md
secure-cookie
==============

[![NPM Version][npm-badge]][npm-url]
[![Build Status][ci-badge]][ci-url]
[![Test Coverage][codecov-badge]][codecov-url]
[![Licencse][license-badge]][license-url]

Nodejs cookie library with signing and encryption support. For those familiar with
[`cookies`](https://github.com/pillarjs/cookies)
this library is almost the same plus encryption and additional configuration support.

## Installation

This library is published in the NPM registry and can be installed using any compatible package manager.

```sh
npm install secure-cookie --save

# For Yarn, use the command below.
yarn add secure-cookie
```

## Documentation

### Signed Cookies

```javascript
const {Cookies, KeyStore} = require('secure-cookies')

const app = express()
app.use(Cookies.express({
  signed: true,
  keyStore: new KeyStore({
    signing: {
  //  encoding: 'base64',
  //  algorithm: 'sha1',
      keys: ["mysigningkey"]
    }
  })
}))

//In a route handler
app.get('/some-route', function (req, res, next) {
  //This will create a cookie named MC with given value.
  // Because of signing is enabled, a new cookie with MC.sig will also be created
  // and would contain signature of the cookie.
  req.cookies.set('MC', "someValue")
})
```

### Encrypted Cookies
```javascript
const {Cookies, KeyStore} = require('secure-cookies')

const app = express()

app.use(Cookies.express({
  signed: true,
  keyStore: new KeyStore({
    encryption: {
   // algorithm: 'aes-192-ccm',
   // authTagLength: 16,
   // encoding: 'hex',
      keys: ["a24bytesecretmustchanged"]
    }
  })
}))

app.get('/set-cookie', function (req, res, next) {
  //This will create a cookie named MC with and with its encrtypted value.
  req.cookies.set('MC', "someValue")
})
app.get('/get-cookie', function (req, res, next) {
  // get decrypted value without hassle
  const myCookie = req.cookies.get('MC')
  assert.equal(myCookie, "someValue")
})
```

Make sure selected algorithm is supported by your NodeJs version.
By default `aes-192-ccm` is selected. You can override that and related settings from KeyStore constructor options.
If the algorithm you would like to use is missing from the default ones you can add it by following:

```javascript
const {KeyStore} = require('secure-cookies')
KeyStore.cipherInfo['aes-xxx-xxx'] =  { ivLength: 16, keyLength: 16 }
```

You can see included algorithms from [src/ciphers.ts](/src/ciphers.ts)

For all options and internals have a look at to [API documentation](./docs/README.md).

## License

Released under [MIT License](./LICENSE).


[npm-badge]: https://img.shields.io/npm/v/secure-cookie.svg
[npm-url]: https://www.npmjs.com/package/secure-cookie
[codecov-badge]: https://img.shields.io/codecov/c/gh/ayZagen/secure-cookie?logo=codecov
[codecov-url]: https://codecov.io/gh/ayZagen/secure-cookie
[ci-badge]: https://github.com/ayZagen/secure-cookie/workflows/ci/badge.svg
[ci-url]: https://github.com/ayZagen/secure-cookie/actions?query=workflow%3Aci
[license-badge]: https://badgen.net/github/license/ayZagen/secure-cookie
[license-url]: ./LICENSE


CHANGELOG.md
# [0.1.0](https://github.com/ayZagen/secure-cookie/compare/v0.0.7...v0.1.0) (2021-09-23)


### Features

* allow empty keystore constructor params ([1cfbf0a](https://github.com/ayZagen/secure-cookie/commit/1cfbf0a5ad5d85f4a81d290340cfc0efdb28beb8))

## [0.0.7](https://github.com/ayZagen/secure-cookie/compare/v0.0.6...v0.0.7) (2021-09-21)


### Bug Fixes

* koa middleware fails with async ([e57275f](https://github.com/ayZagen/secure-cookie/commit/e57275fe2bff66c853a6ba146e10341059715916))

## [0.0.6](https://github.com/ayZagen/secure-cookie/compare/v0.0.5...v0.0.6) (2021-09-21)

## [0.0.5](https://github.com/ayZagen/secure-cookie/compare/v0.0.4...v0.0.5) (2021-09-21)

## [0.0.4](https://github.com/ayZagen/secure-cookie/compare/v0.0.3...v0.0.4) (2021-09-21)

## [0.0.3](https://github.com/ayZagen/secure-cookie/compare/v0.0.2...v0.0.3) (2021-09-21)

## [0.0.2](https://github.com/ayZagen/secure-cookie/compare/v0.0.1...v0.0.2) (2021-09-20)


### Bug Fixes

* padding ignored for encryption without iv ([703af9a](https://github.com/ayZagen/secure-cookie/commit/703af9aa9e2aaa5d5df96a832a2360e4c3a738fd))

## 0.0.1 (2021-09-20)



src/keystore.ts
/* eslint-disable @typescript-eslint/ban-ts-comment */
import compare from 'tsscmp'

import crypto, {BinaryToTextEncoding, CipherKey, Encoding} from "crypto";
import {CIPHER_INFO} from "./ciphers";

const AUTH_TAG_REQUIRED = /-(gcm|ccm)/

export interface KeyStoreOpts {
  signing?: {
    keys: string[],
    algorithm?: string | 'blake2b512' | 'blake2s256' | 'gost' | 'md4' | 'md5' | 'rmd160' | 'sha1' | 'sha224'
      | 'sha256' | 'sha3-224' | 'sha3-256' | 'sha3-384' | 'sha3-512' | 'sha384' | 'sha512' | 'sha512-224'
      | 'sha512-256' | 'shake128' | 'shake256' | 'sm3';
    encoding?: BinaryToTextEncoding
  },
  encryption?: {
    keys: string[] | CipherKey[],
    algorithm?: keyof typeof CIPHER_INFO | string,
    encoding?: Encoding
    authTagLength?: number
  }
}

export type EncryptOptions = Required<KeyStoreOpts['encryption']> & { key?: string | undefined }
export type DecryptOptions = Required<KeyStoreOpts['encryption']> & {
  key?: string | CipherKey | undefined,
  iv?: string | Buffer | undefined,
  authTag?: string | Buffer | undefined
}

export class KeyStore {
  encryption: Required<NonNullable<KeyStoreOpts['encryption']>>;
  signing: Required<NonNullable<KeyStoreOpts['signing']>>;

  static cipherInfo = CIPHER_INFO as Record<string | keyof typeof CIPHER_INFO, {
    ivLength: number | undefined,
    keyLength: number
  }>

  constructor(opts?: KeyStoreOpts) {
    opts = opts || {}

    this.encryption = Object.assign({
      algorithm: 'aes-192-ccm',
      authTagLength: 16,
      encoding: 'hex',
      keys: []
    }, opts.encryption || {} as any)

    this.signing = Object.assign({
      encoding: 'base64',
      algorithm: 'sha1',
      keys: []
    }, opts.signing || {} as any)

  }

  encrypt(data?: null, options?: Partial<EncryptOptions>): null
  encrypt(data: string | Buffer, options?: Partial<EncryptOptions>): string
  encrypt(data?: string | Buffer | null, options?: Partial<EncryptOptions>): string | null {
    if (!data) {
      return null
    }
    const {
      keys,
      algorithm,
      encoding,
      authTagLength,
      key
    }: EncryptOptions = options ? Object.assign({}, this.encryption, options) : this.encryption

    const secret = key || keys[0]

    if (!secret) {
      throw new Error("no key found")
    }

    const cipherInfo = KeyStore.cipherInfo[algorithm]
    if (!cipherInfo) {
      throw new Error("unsupported cipher")
    }

    const iv = cipherInfo.ivLength ? crypto.randomBytes(cipherInfo.ivLength) : null;
    const dataBuff = typeof data === "string" ? Buffer.from(data, 'utf-8') : data

    const cipher = crypto.createCipheriv(algorithm as any, secret, iv, {authTagLength})

    const text = cipher.update(dataBuff);
    const pad = cipher.final();
    let authTag: Buffer | undefined;
    if (AUTH_TAG_REQUIRED.test(algorithm)) {
      authTag = cipher.getAuthTag();
    }

    return Buffer.concat([
      ...iv ? [iv] : [],
      ...authTag ? [authTag] : [],
      text,
      pad
    ]).toString(encoding)
  }

  decrypt(data?: null, options?: Partial<DecryptOptions>): null
  decrypt(data: string | Buffer, options?: Partial<DecryptOptions>): string
  decrypt(data?: string | Buffer | null, options?: Partial<DecryptOptions>): string | null {
    if (!data) {
      return null
    }

    const finalOptions: DecryptOptions = options ? Object.assign({}, this.encryption, options) : this.encryption

    const {
      encoding,
      key,
      keys: defaultKeys,
      algorithm,
      authTagLength,
    } = finalOptions

    const keys = key ? [key] : defaultKeys
    if (keys.length === 0) {
      throw new Error("keys required for encrypted cookies")
    }

    let {
      iv,
      authTag
    } = finalOptions

    let dataBuff = typeof data === "string" ? Buffer.from(data, encoding) : data

    const cipherInfo = KeyStore.cipherInfo[algorithm]
    if (!cipherInfo) {
      throw new Error("unsupported cipher")
    }

    if (typeof iv === "string") {
      iv = Buffer.from(iv, encoding)
    }

    if (typeof authTag === "string") {
      authTag = Buffer.from(authTag, encoding)
    }

    if (cipherInfo.ivLength !== undefined) {
      if (!iv) {
        iv = dataBuff.slice(0, cipherInfo.ivLength)
      }
      dataBuff = dataBuff.slice(cipherInfo.ivLength, dataBuff.length)
    }


    if (AUTH_TAG_REQUIRED.test(algorithm)) {
      if (!authTag) {
        authTag = dataBuff.slice(0, authTagLength)
      }
      dataBuff = dataBuff.slice(authTagLength, dataBuff.length)
    }

    for (let i = 0; i < keys.length; i++) {
      const message = KeyStore.doDecrypt(dataBuff, {...finalOptions, key: keys[i], iv, authTag,});
      if (message !== null) return message
    }
    return null
  }

  private static doDecrypt(data: Buffer, options: DecryptOptions): string | null {
    const {algorithm, key, iv, authTagLength, authTag} = options
    const decipher = crypto.createDecipheriv(algorithm as any, key!, iv as Buffer || null, {authTagLength});

    if (authTag) {
      decipher.setAuthTag(authTag as Buffer)
    }

    const plainText = decipher.update(data)
    let final: Buffer
    try {
      final = decipher.final()
    } catch(e:any) {
      // authentication failed
      return null
    }
    return Buffer.concat([plainText, final]).toString('utf-8')
  }

  //region: signing

  sign(data?: null, key?: string | CipherKey): null
  sign(data: string, key?: string | CipherKey): string
  sign(data?: string | null, key?: string | CipherKey): string | null {
    if (!data) {
      return null
    }
    const {algorithm, encoding, keys} = this.signing
    key = key || keys[0] as CipherKey
    return crypto
      .createHmac(algorithm, key)
      .update(data).digest(encoding)
      .replace(/\/|\+|=/g, function (x) {
        return ({"/": "_", "+": "-", "=": ""})[x] as string
      })
  }

  verify(data: string, digest: string): boolean {
    return this.indexOf(data, digest) > -1
  }

  indexOf(data: string, digest: string): number {
    const {keys} = this.signing

    if (keys.length === 0) {
      throw new Error("keys required for signed cookies")
    }

    for (let i = 0; i < keys.length; i++) {
      if (compare(digest, this.sign(data, keys[i] as string))) return i
    }
    return -1
  }

  //end-region: signing

}

